File: Clients/C++/Knobs.cpp
//#include "stdafx.h" //Uncomment if you're into that sort of thing
#include "Knobs.h"

#include <winsock2.h>
#include <ws2tcpip.h>
#include <thread>

#pragma comment (lib, "Ws2_32.lib")
#pragma comment (lib, "Mswsock.lib")
#pragma comment (lib, "AdvApi32.lib")

int Knobs::knobs[NUM_KNOBS];

void Knobs::start()
{
	std::thread whatCouldPossiblyGoWrong(Knobs::loop);
	whatCouldPossiblyGoWrong.detach();
}

void Knobs::stop()
{
	//NO BRAKES
}


void Knobs::loop() {
	WSADATA wsaData;
	SOCKET ConnectSocket = INVALID_SOCKET;
	struct addrinfo *result = NULL,
		*ptr = NULL,
		hints;
	char recvbuf[NUM_KNOBS];
	int iResult;
	int recvbuflen = NUM_KNOBS;

	// Initialize Winsock
	iResult = WSAStartup(MAKEWORD(2, 2), &wsaData);
	if (iResult != 0) {
		printf("WSAStartup failed with error: %d\n", iResult);
		return;
	}

	ZeroMemory(&hints, sizeof(hints));
	hints.ai_family = AF_UNSPEC;
	hints.ai_socktype = SOCK_STREAM;
	hints.ai_protocol = IPPROTO_TCP;

	// Resolve the server address and port
	iResult = getaddrinfo(KNOB_HOST, KNOB_PORT, &hints, &result);
	if (iResult != 0) {
		printf("getaddrinfo failed with error: %d\n", iResult);
		WSACleanup();
		return;
	}

	// Attempt to connect to an address until one succeeds
	for (ptr = result; ptr != NULL; ptr = ptr->ai_next) {

		// Create a SOCKET for connecting to server
		ConnectSocket = socket(ptr->ai_family, ptr->ai_socktype,
			ptr->ai_protocol);
		if (ConnectSocket == INVALID_SOCKET) {
			printf("socket failed with error: %ld\n", WSAGetLastError());
			WSACleanup();
			return;
		}

		// Connect to server.
		iResult = connect(ConnectSocket, ptr->ai_addr, (int)ptr->ai_addrlen);
		if (iResult == SOCKET_ERROR) {
			closesocket(ConnectSocket);
			ConnectSocket = INVALID_SOCKET;
			continue;
		}
		break;
	}

	freeaddrinfo(result);

	if (ConnectSocket == INVALID_SOCKET) {
		printf("Unable to connect to server!\n");
		WSACleanup();
		return;
	}

	// Receive until the peer closes the connection
	do {

		iResult = recv(ConnectSocket, recvbuf, recvbuflen, 0);
		if (iResult > 0) {
			for (int i = 0; i < recvbuflen; i++) {
				knobs[i] = recvbuf[i];
			}
		}
		else if (iResult == 0)
			printf("Connection closed\n");
		else
			printf("recv failed with error: %d\n", WSAGetLastError());

	} while (iResult > 0);

	// cleanup
	closesocket(ConnectSocket);
	WSACleanup();
}


float Knobs::get(int index)
{
	return Knobs::get(index, 0, 1);
}

float Knobs::get(int index, float max)
{
	return Knobs::get(index, 0, max); return 0.0f;
}

float Knobs::get(int index, float min, float max)
{
	float value = (float)Knobs::knobs[index] / 127.0f;
	return (1 - value) * min + value * max;
}

File: Clients/C#/Knobs.cs
using System;
using System.Net.Sockets;
using System.Threading;

public class Knobs
{
    private static int size;
    public static int[] knobs;

    private static Socket sock;

    private static Thread worker;
    private static bool running = false;

    public static void Start(int numKnobs)
    {
        if (!running)
        {
            size = numKnobs;
            knobs = new int[size];

            try
            {
                sock = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
                sock.Connect("localhost", 8008);

                worker = new Thread(new ThreadStart(readFromServer));
                worker.Start();

                running = true;
            }
            catch (SocketException e)
            {
                Console.WriteLine(e);
            }
        }
    }

    public static void Stop()
    {
        if (running)
        {
            sock.Shutdown(SocketShutdown.Both);
            sock.Close();
            running = false;
        }
    }

    public static float Get(int index)
    {
        return get(index, 0, 1);
    }

    public static float Get(int index, float max)
    {
        return get(index, 0, max);
    }

    public static float Get(int index, float min, float max)
    {
        float value = (float) knobs[index] / 127.0f;
        return (1 - value) * min + value * max;
    }


    private static void ReadFromServer()
    {
        byte[] sockBuffer = new byte[size];

        try
        {
            while (running)
            {
                if (sock.Receive(sockBuffer, size, SocketFlags.None) > 0)
                {
                    for (int i = 0; i < size; i++)
                    {
                        knobs[i] = sockBuffer[i];
                    }
                }
                else
                {
                    Console.WriteLine("Socket connection lost; aborting");
                    sock.Shutdown(SocketShutdown.Both);
                    sock.Close();
                    break;
                }
            }
        }
        catch (SocketException e) {
            Console.WriteLine(e);
        }
    }
}